File: typescript/packages/wallets/crossmint/src/mint.ts
import type { CrossmintApiClient } from "@crossmint/common-sdk-base";
import type { Plugin, WalletClient } from "@goat-sdk/core";
import { z } from "zod";
import { getCrossmintChainString, isChainSupportedByMinting } from "./chains";

export const mintingFactory = (client: CrossmintApiClient): (() => Plugin<WalletClient>) => {
    return () => ({
        name: "minting",
        supportsSmartWallets: () => true,
        supportsChain: (chain) => {
            if (chain.type === "evm") {
                return isChainSupportedByMinting(chain.id ?? 0);
            }

            if (chain.type === "aptos" || chain.type === "solana") {
                return true;
            }

            return false;
        },
        getTools: async (walletClient: WalletClient) => {
            return [
                {
                    name: "create_nft_collection",
                    description: "This {{tool}} creates an NFT collection and returns the ID of the collection.",
                    parameters: createCollectionParametersSchema,
                    method: createCollectionMethod(client, walletClient),
                },
                {
                    name: "get_all_collections",
                    description: "This {{tool}} gets all the collections created by the user.",
                    parameters: getAllCollectionsParametersSchema,
                    method: getAllCollectionsMethod(client, walletClient),
                },
                {
                    name: "mint_nft",
                    description:
                        "This {{tool}} mints an NFT to a recipient from a collection and returns the transaction hash. Requires a collection ID of an already deployed collection.",
                    parameters: mintNFTParametersSchema,
                    method: mintNFTMethod(client, walletClient),
                },
            ];
        },
    });
};

const createCollectionParametersSchema = z.object({
    metadata: z
        .object({
            name: z.string().describe("The name of the collection"),
            description: z.string().describe("A description of the NFT collection"),
            image: z.string().optional().describe("URL pointing to an image that represents the collection"),
            symbol: z
                .string()
                .optional()
                .describe("Shorthand identifier for the NFT collection (Max length: 10). Defaults to 'TOKEN'"),
        })
        .default({
            name: "My first Minting API Collection",
            description:
                "An NFT Collection created with the Crossmint Minting API - learn more at https://www.crossmint.com/products/nft-minting-api",
            image: "https://www.crossmint.com/assets/crossmint/logo.png",
        })
        .describe("The metadata of the collection"),
    fungibility: z
        .enum(["semi-fungible", "non-fungible"])
        .optional()
        .default("non-fungible")
        .describe("Whether or not this collection is fungible (e.g ERC-1155 vs ERC-721)"),
    transferable: z
        .boolean()
        .optional()
        .default(true)
        .describe("Whether or not the NFTs in this collection are transferable"),
});

const getAllCollectionsParametersSchema = z.object({});

const mintNFTParametersSchema = z.object({
    collectionId: z.string().describe("The ID of the collection to mint the NFT in"),
    recipient: z
        .string()
        .describe(
            "A locator for the recipient of the NFT, in the format `<address>` if it's a wallet or `email:<email_address>` if it's an email",
        ),
    metadata: z
        .object({
            name: z.string().describe("The name of the NFT"),
            description: z.string().max(64).describe("The description of the NFT"),
            image: z.string().describe("URL pointing to the NFT image"),
            animation_url: z.string().optional().describe("URL pointing to the NFT animation"),
            attributes: z
                .array(
                    z.object({
                        display_type: z
                            .enum(["number", "boost_number", "boost_percentage"])
                            .describe("The type of the attribute, if it's a number or a percentage"),
                        value: z.string().describe("The trait value"),
                    }),
                )
                .optional()
                .describe("The attributes of the NFT"),
        })
        .describe("The metadata of the NFT"),
});

function getAllCollectionsMethod(client: CrossmintApiClient, walletClient: WalletClient) {
    return async () => {
        const response = await fetch(`${client.baseUrl}/collections/`, {
            headers: client.authHeaders,
        });

        return await response.json();
    };
}

function createCollectionMethod(client: CrossmintApiClient, walletClient: WalletClient) {
    return async (parameters: z.infer<typeof createCollectionParametersSchema>) => {
        const response = await fetch(`${client.baseUrl}/api/2022-06-09/collections/`, {
            method: "POST",
            body: JSON.stringify({
                ...parameters,
                chain: getCrossmintChainString(walletClient.getChain()),
            }),
            headers: client.authHeaders,
        });

        const result = await response.json();

        if (result.error) {
            throw new Error(result.message);
        }

        const { id, actionId } = result;

        const action = await waitForAction(actionId, client);

        const chain = getCrossmintChainString(walletClient.getChain());

        return {
            collectionId: id,
            chain,
            contractAddress: action.data.collection.contractAddress,
        };
    };
}

function mintNFTMethod(client: CrossmintApiClient, walletClient: WalletClient) {
    return async (parameters: z.infer<typeof mintNFTParametersSchema>) => {
        let recipient: string;

        if (parameters.recipient.startsWith("email:")) {
            recipient = `${parameters.recipient}:${getCrossmintChainString(walletClient.getChain())}`;
        } else {
            recipient = `${getCrossmintChainString(walletClient.getChain())}:${parameters.recipient}`;
        }

        const response = await fetch(`${client.baseUrl}/api/2022-06-09/collections/${parameters.collectionId}/nfts`, {
            method: "POST",
            body: JSON.stringify({
                recipient,
                metadata: parameters.metadata,
            }),
            headers: {
                ...client.authHeaders,
                "Content-Type": "application/json",
            },
        });

        const result = await response.json();

        if (result.error) {
            throw new Error(result.message);
        }

        const { id, actionId, onChain } = result;

        const action = await waitForAction(actionId, client);

        return {
            id: id,
            collectionId: parameters.collectionId,
            contractAddress: onChain.contractAddress,
            chain: action.data.chain,
        };
    };
}

async function waitForAction(actionId: string, client: CrossmintApiClient) {
    let attempts = 0;
    while (true) {
        attempts++;
        const response = await fetch(`${client.baseUrl}/api/2022-06-09/actions/${actionId}`, {
            headers: client.authHeaders,
        });

        const body = await response.json();

        if (response.status === 200 && body.status === "succeeded") {
            return body;
        }
        await new Promise((resolve) => setTimeout(resolve, 1000));
        if (attempts >= 60) {
            throw new Error(`Timed out waiting for action ${actionId} after ${attempts} attempts`);
        }
    }
}

File: python/src/wallets/crossmint/goat_wallets/crossmint/mint.py
from goat.decorators.tool import Tool
from goat_wallets.evm import EVMWalletClient
from typing import Dict, Any, List

from .parameters import CollectionParameters, MintNFTParameters, EmptyParameters
from .api_client import CrossmintWalletsAPI
from .chains import get_crossmint_chain_string, is_story_chain


class CrossmintMintService:
    def __init__(self, api_client: CrossmintWalletsAPI):
        self.api_client = api_client

    def _get_endpoint_root(self, chain: str) -> str:
        """Get the appropriate endpoint root based on chain type.
        
        Args:
            chain: Chain identifier
            
        Returns:
            str: API endpoint root path
        """
        return "/api/v1/ip" if is_story_chain(chain) else "/api/2022-06-09"

    @Tool({
        "description": "Create a new NFT collection",
        "parameters_schema": CollectionParameters
    })
    def create_collection(self, wallet_client: EVMWalletClient, parameters: dict) -> dict:
        try:
            chain = get_crossmint_chain_string(wallet_client.get_chain())
            
            result = self.api_client.create_collection(parameters, chain)
            if result.get("error"):
                raise Exception(result["message"])
            
            action = self.api_client.wait_for_action(result["actionId"])
            
            return {
                "collectionId": result["id"],
                "chain": chain,
                "contractAddress": action["data"]["collection"]["contractAddress"]
            }
        except Exception as error:
            raise Exception(f"Failed to create collection: {error}")

    @Tool({
        "description": "Get all collections",
        "parameters_schema": EmptyParameters
    })
    def get_all_collections(self, wallet_client: EVMWalletClient, parameters: dict) -> List[Dict[str, Any]]:
        try:
            chain = get_crossmint_chain_string(wallet_client.get_chain())
            collections = self.api_client.get_all_collections(chain)
            return list(collections.values()) if isinstance(collections, dict) else collections
        except Exception as error:
            raise Exception(f"Failed to get collections: {error}")

    @Tool({
        "description": "Mint an NFT in a collection",
        "parameters_schema": MintNFTParameters
    })
    def mint_nft(self, wallet_client: EVMWalletClient, parameters: dict) -> dict:
        try:
            chain = get_crossmint_chain_string(wallet_client.get_chain())
            
            # Format recipient appropriately based on chain and recipient type
            recipient = f"email:{parameters['recipient']}:{chain}" if parameters["recipient_type"] == "email" else f"{chain}:{parameters['recipient']}"
            
            result = self.api_client.mint_nft(
                parameters["collection_id"],
                recipient,
                parameters["metadata"],
                chain
            )
            
            if result.get("error"):
                raise Exception(result["message"])
            
            action = self.api_client.wait_for_action(result["actionId"])
            
            return {
                "id": result["id"],
                "collectionId": parameters["collection_id"],
                "contractAddress": result["onChain"]["contractAddress"],
                "chain": action["data"]["chain"]
            }
        except Exception as error:
            raise Exception(f"Failed to mint NFT: {error}")
